README

				Net::Dict
				  v2.08

This distribution contains the Net::Dict module for Perl.
Net::Dict is a class implementing a simple client API
for the DICT protocol defined in RFC2229.

This release is the first for a number of years. The main purpose of this
release is to get the testsuite passing: the testsuite refers to a live DICT
server, and the number of dictionaries has grown since I last released the module
and the dedicated test server is no longer provided.

To install this module, you should just have to run the following:

	% perl Makefile.PL
	% make
	% make test
	% make install

When you run "perl Makefile.PL" you'll be asked for the hostname
and port for the DICT server used when testing. If you're not
going to run "make install", then just press return. You should
be able to just press return on the two questions anyway.

This module now supports the AUTH optional capability. To use this
you will need the Digest::MD5 module, available from CPAN.

The module is documented using pod. When you "make install", you
will get a man-page Net::Dict. You can also generate HTML using pod2html:

	% pod2html lib/Net/Dict.pm

Three sample clients are included in this distribution.
Any additional modules required are noted, and available from CPAN.

    dict
	A basic command-line client, based on the C dict client
	by Rik Faith.
	Requires: AppConfig, AppConfig::Std

    tkdict
	A first cut at a Perl/Tk client. This is pretty rough;
	any suggestions or patches are welcome!
	Requires: AppConfig, AppConfig::Std, Tk, Tk::Dialog

    examples/simple.pl
	Illustrates basic use of Net::Dict.

Net::Dict was written by Dmitry Rubinstein, but is now maintained by me.


Neil Bowers <neil@bowers.com>


Changes

		ChangeLog for Net-Dict distribution

2.12 2013-11-18

    - Corrected the dependency I meant to add in the previous release.
      I added a dependency on Net::Dict (ie itself) instead.

2.11 2013-11-15

    - Added missing dependency (AppConfig::Std) to Makefile.PL

2.10 2013-07-20

    - Reformatted this file according to CPAN::Changes::Spec
    - Repository details added to metadata (Makefile.PL) and pod
    - License type added to metadata (Makefile.PL)

2.09 2011-12-18

    - Fixed tests that started breaking due to changes in the dict.org server
    - Renamed ChangeLog to Changes & tweaked formatting to CPAN::Changes::Spec

2.08 2011-08-02

    - updated testsuite to refer to dict.org, as test.dict.org no longer exists
    - updated testsuite to reflect the much longer list of databases now hosted on dict.org

2.07 2003-05-06

	- updated testsuite to refer to test.dict.org,
	  and to reflect changes in the databases.

2.06 2002-03-23

	- imported into my home machine's CVS repository
	- updated email address

2.05 2001-04-25

	- moved the inline documentation to a separate file Dict.pod
	- added examples/portuguese.pl which illustrates accessing
	  an english-portuguese dictionary.
	  Example from Jose Joao Dias de Almeida <jj@di.uminho.pt>.

2.04 2001-04-23

	- tidied up the code for auth(), removing debugging statements, etc.
	- added documentation for the auth() method.
	- renamed auth.t to auth.test - don't want this run as
	  part of "make test": it needs my local config for testing.
	  Do something about that later.

2.03 2001-04-23

	- Added code which parses the welcome banner, to get msg id and
	  optional capabilities.
	- Added capabilities() method which returns a list of
	  supported optional capabilities.
	- Added has_capability() method for checking whether a
	  capability is supported by the server.
	- msg_id() method which returns the msg id from the server.
	  This is used in the auth() method.
	- Added auth() method, which uses Digest::MD5.
	- Created a testsuite for auth - auth.t


2.02 2001-04-03

	- Oops - forgot to add documentation for the status() method.

2.01 2001-04-03

	- Added status() method to Net::Dict - returns the string
	  returned by the DICT server when STATUS command is sent.
	  Couple of test cases in t/connection.t

	- When using the sample dict client, if no definition was
	  found, then it will use Levenshtein or Soundex matching
	  to look for close words. If the server doesn't support
	  either strategy, then it just gives a basic error message.

	- Updated the testsuite - new databases on dict.org meant
	  that certain tests failed (eg where the date is included
	  in the title of a database).

2.00 2001-04-01

	- up'd the major version number - this will be the first public
	  release version since changing the API for the constructor.
		- updated dict and tkdict to use the new method name

	- Various documentation updates, including:
		- adding more to the descriptive section of the documentation.
		- reformatting the METHODS section

	- strats() method renamed to strategies(). The old name is
	  retained for backwards compatibility.

	- Put a hack in the match.t test to supress unwanted output
	  from _print_isa function in Net::Cmd.

	- Removed the dependence on Net::Config from Makefile.PL


1.09 2001-03-26

	- Send the CLIENT command to identify us before any other command
	  is sent.

	- Don't need to "use Net::Config" now

	- dbTitle() checks whether the given DB name is valid.
	  If it isn't, and debug is set to non-zero, then we now carp.

	- Fixed a bug in define() - couldn't handle multi-word entries, eg:
		$dict->define("oboe d'amore");
	  didn't work as it should. The private _DEFINE method now quotes
	  all arguments before passing them on, since having everything
	  quoted is ok by RFC 2229.

	- Fixed the same bug in match() method.

	- Finished first pass at testsuite for define() method.

1.08 2001-03-22

	- first version of testsuite - not the full set, but enough
	  to get a few people to test and find out if it's sensible.
	- Makefile.PL updated to get hostname and port for test server,
	  it builds a config file in t/

	- dbInfo now returns a string rather than an array of lines.
	  This means it now matches the documentation!

	- dbTitle() returns undef if you request a title of a
	  non-existent database.

 	- Now checks for legality of arg names passed to constructor
	- constructor requires hostname as first argument
	- don't look for default list of hosts to try from Net::Config
	- updated checking of arguments to constructor and error messages
	- changed all self variables from $obj to $self
	- improved wording of error messages when checking method arg lists
	- private method _CLIENT now takes arg, rather than hard-coding
	  reference to package variable $CLIENT_INFO
	- Removed references in to the doc to ConfigFile and HTML
	  arguments - they weren't actually supported - now mention
	  this in the LIMITATIONS section
	- Put an example of use of constructor with all arguments
	  in the doc

1.07 2001-03-04

	- Updated the one-line description in the NAME pod section.
	  Previous one was a bit terse - that's what shows up
	  on search.cpan.org, and similar places.

1.06 2001-03-04

	- created tkdict, first cut at a Perl/Tk DICT client.
	  The interface is currently very DICT protocol centric.
	- added dbTitle() method, which is used to query the title
	  string for a specific database.
	- the description strings returned by dbs() and strats() were
	  quoted with double strings (if that's what the server returned).
	  Similarly every word returned by match() was quoted.
	  Now the quotation marks are removed.

1.05 2001-03-01
	- added "dict", a sample client script
	- strats() method was including a newline in the description
	  of each strategy, unlike dbs(), which chomp()s the description.
	  strats() now chomps as well!
	- added Client option to Net::Dict, for CLIENT identifier string

lib/Net/Dict.pm
#
# Net::Dict.pm
#
# Copyright (C) 2001-2003 Neil Bowers <neil@bowers.com>
# Copyright (c) 1998 Dmitry Rubinstein <dimrub@wisdom.weizmann.ac.il>.
#
# All rights reserved.  This program is free software; you can
# redistribute it and/or modify it under the same terms as Perl
# itself.
#

package Net::Dict;

use strict;
use IO::Socket;
use Net::Cmd;
use Carp;

use vars qw(@ISA $debug);
our $VERSION = '2.12';

#-----------------------------------------------------------------------
# Default values for arguments to new(). We also use this to
# determine valid argument names - if it's not a key of this hash,
# then it's not a valid argument.
#-----------------------------------------------------------------------
my %ARG_DEFAULT =
(
 Port    => 2628,
 Timeout => 120,
 Debug   => 0,
 Client  => "Net::Dict v$VERSION",
);

@ISA = qw(Net::Cmd IO::Socket::INET);

#=======================================================================
#
# new()
#
# constructor - open connection to host, get a list of databases,
# and send CLIENT identification command.
#
#=======================================================================
sub new
{
    @_ > 1 or croak 'usage: Net::Dict->new() takes at least a HOST name';
    my $class  = shift;
    my $host   = shift;
    int(@_) % 2 == 0 or croak 'Net::Dict->new(): odd number of arguments';
    my %inargs = @_;

    my $self;
    my $argref;


    return undef unless defined $host;

    #-------------------------------------------------------------------
    # Process arguments, setting defaults if needed
    #-------------------------------------------------------------------
    $argref = {};
    foreach my $arg (keys %ARG_DEFAULT)
    {
        $argref->{$arg} = exists $inargs{$arg}
                          ? $inargs{$arg}
                          : $ARG_DEFAULT{$arg};
        delete $inargs{$arg};
    }
    if (keys(%inargs) > 0)
    {
        croak "Net::Dict->new(): unknown argument - ",
            join(', ', keys %inargs);
    }

    #-------------------------------------------------------------------
    # Make the connection
    #-------------------------------------------------------------------
    $self = $class->SUPER::new(PeerAddr => $host,
                               PeerPort => $argref->{Port},
                               Proto    => 'tcp',
                               Timeout  => $argref->{Timeout}
                               );

    return undef
	unless defined $self;

    ${*$self}{'net_dict_host'} = $host;

    $self->autoflush(1);
    $self->debug($argref->{Debug});

    if ($self->response() != CMD_OK)
    {
        $self->close();
        return undef;
    }

    # parse the initial 220 response
    $self->_parse_banner($self->message);

    #-------------------------------------------------------------------
    # Send the CLIENT command which identifies the connecting client
    #-------------------------------------------------------------------
    $self->_CLIENT($argref->{Client});

    #-------------------------------------------------------------------
    # The default - search ALL dictionaries
    #-------------------------------------------------------------------
    $self->setDicts('*');

    return $self;
}

sub dbs
{
    @_ == 1 or croak 'usage: $dict->dbs() - takes no arguments';
    my $self = shift;


    $self->_get_database_list();
    return %{${*$self}{'net_dict_dbs'}};
}

sub setDicts
{
    my $self = shift;

    @{${*$self}{'net_dict_userdbs'}} = @_;
}

sub serverInfo
{
    @_ == 1 or croak 'usage: $dict->serverInfo()';
    my $self = shift;

    return 0
        unless $self->_SHOW_SERVER();
    my $info = join('', @{$self->read_until_dot});
    $self->getline();
    $info;
}

sub dbInfo
{
    @_ == 2 or croak 'usage: $dict->dbInfo($dbname) - one argument only';
    my $self = shift;

    if ($self->_SHOW_INFO(@_))
    {
        return join('', @{$self->read_until_dot()});
    }
    else
    {
        return undef;
    }
}

sub dbTitle
{
    @_ == 2 or croak 'dbTitle() method expects one argument - DB name';
    my $self   = shift;
    my $dbname = shift;


    $self->_get_database_list();
    if (exists ${${*$self}{'net_dict_dbs'}}{$dbname})
    {
        return ${${*$self}{'net_dict_dbs'}}{$dbname};
    }
    else
    {
        carp 'dbTitle(): unknown database name' if $self->debug;
        return undef;
    }
}

sub strategies
{
    @_ == 1 or croak 'usage: $dict->strategies()';
    my $self = shift;
    return 0
        unless $self->_SHOW_STRAT();
    my(%strats, $name, $desc);
    foreach (@{$self->read_until_dot()})
    {
        ($name, $desc) = (split /\s/, $_, 2);
        chomp $desc;
        $strats{$name} = _unquote($desc);
    }
    $self->getline();
    %strats;
}

sub define
{
    @_ >= 2 or croak 'usage: $dict->define($word [, @dbs]) - takes at least one argument';
    my $self = shift;
    my $word = shift;
    my @dbs = (@_ > 0) ? @_ : @{${*$self}{'net_dict_userdbs'}};
    croak 'select some dictionaries with setDicts or supply as argument to define'
        unless @dbs;
    my($db, @defs);


    #-------------------------------------------------------------------
    # check whether we got an empty word
    #-------------------------------------------------------------------
    if (!defined($word) || $word eq '')
    {
        carp "empty word passed to define() method";
        return undef;
    }

    foreach $db (@dbs)
    {
        next
            unless $self->_DEFINE($db, $word);

        my ($defNum) = ($self->message =~ /^\d{3} (\d+) /);
        foreach (0..$defNum-1)
        {
            my ($d) = ($self->getline =~ /^\d{3} ".*" (\w+) /);
            my ($def) = join '', @{$self->read_until_dot};
            push @defs, [$d, $def];
        }
        $self->getline();
    }
    \@defs;
}

sub match
{
    @_ >= 3 or croak 'usage: $self->match($word, $strat [, @dbs]) - takes at least two arguments';
    my $self = shift;
    my $word = shift;
    my $strat = shift;
    my @dbs = (@_ > 0) ? @_ : @{${*$self}{'net_dict_userdbs'}};
    croak 'define some dictionaries by setDicts or supply as argument to define'
        unless @dbs;
    my ($db, @matches);

    #-------------------------------------------------------------------
    # check whether we got an empty pattern
    #-------------------------------------------------------------------
    if (!defined($word) || $word eq '')
    {
        carp "empty pattern passed to match() method";
        return undef;
    }

    foreach $db (@dbs)
    {
        next unless $self->_MATCH($db, $strat, $word);

        my ($db, $w);
        foreach (@{$self->read_until_dot}) {
            ($db, $w) = split /\s/, $_, 2;
            chomp $w;
            push @matches, [$db, _unquote($w)];
        }
        $self->getline();
    }
    \@matches; 
}

sub auth
{
    @_ == 3 or croak 'usage: $dict->auth() - takes two arguments';
    my $self        = shift;
    my $user        = shift;
    my $pass_phrase = shift;
    my $auth_string;
    my $string;
    my $ctx;


    require Digest::MD5;
    $string = $self->msg_id().$pass_phrase;
    $auth_string = Digest::MD5::md5_hex($string);

    if ($self->_AUTH($user, $auth_string))
    {
        #---------------------------------------------------------------
        # clear the cache of database names
        # next time a method needs them, this will cause us to go
        # back to the server, and thus pick up any AUTH-restricted DBs
        #---------------------------------------------------------------
        delete ${*$self}{'net_dict_dbs'};
    }
    else
    {
        carp "auth() failed with error code ".$self->code() if $self->debug();
        return;
    }
}

sub status
{
    @_ == 1 or croak 'usage: $dict->status() - takes no arguments';
    my $self = shift;
    my $message;


    $self->_STATUS() || return 0;
    chomp($message = $self->message);
    $message =~ s/^\d{3} //;
    return $message;
}

sub capabilities
{
    @_ == 1 or croak 'usage: $dict->capabilities() - takes no arguments';
    my $self = shift;


    return @{ ${*$self}{'net_dict_capabilities'} };
}

sub has_capability
{
    @_ == 2 or croak 'usage: $dict->has_capability() - takes one argument';
    my $self = shift;
    my $cap  = shift;


    return grep(lc($cap) eq $_, $self->capabilities());
}

sub msg_id
{
    @_ == 1 or croak 'usage: $dict->msg_id() - takes no arguments';
    my $self = shift;


    return ${*$self}{'net_dict_msgid'};
}


sub _DEFINE { shift->command('DEFINE', map { '"'.$_.'"' } @_)->response() == CMD_INFO }
sub _MATCH { shift->command('MATCH', map { '"'.$_.'"' } @_)->response() == CMD_INFO }
sub _SHOW_DB { shift->command('SHOW DB')->response() == CMD_INFO }
sub _SHOW_STRAT { shift->command('SHOW STRAT')->response() == CMD_INFO }
sub _SHOW_INFO { shift->command('SHOW INFO', @_)->response() == CMD_INFO }
sub _SHOW_SERVER { shift->command('SHOW SERVER')->response() == CMD_INFO }
sub _CLIENT { shift->command('CLIENT', @_)->response() == CMD_OK }
sub _STATUS { shift->command('STATUS')->response() == CMD_OK }
sub _HELP { shift->command('HELP')->response() == CMD_INFO }
sub _QUIT { shift->command('QUIT')->response() == CMD_OK }
sub _OPTION_MIME { shift->command('OPTION MIME')->response() == CMD_OK }
sub _AUTH { shift->command('AUTH', @_)->response() == CMD_OK }
sub _SASLAUTH { shift->command('SASLAUTH', @_)->response() == CMD_OK }
sub _SASLRESP { shift->command('SASLRESP', @_)->response() == CMD_OK }

sub quit
{
    my $self = shift;

    $self->_QUIT;
    $self->close;
}

sub DESTROY
{
    my $self = shift;

    if (defined fileno($self)) {
        $self->quit;
    }
}

sub response
{
    my $self = shift;
    my $str = $self->getline() || return undef;


    if ($self->debug)
    {
        $self->debug_print(0,$str);
    }

    my($code) = ($str =~ /^(\d+) /);

    ${*$self}{'net_cmd_resp'} = [ $str ];
    ${*$self}{'net_cmd_code'} = $code;

    substr($code,0,1);
}

#=======================================================================
#
# _unquote
#
# Private function used to remove quotation marks from around
# a string.
#
#=======================================================================
sub _unquote
{
    my $string = shift;


    if ($string =~ /^"/)
    {
        $string =~ s/^"//;
        $string =~ s/"$//;
    }
    return $string;
}

#=======================================================================
#
# _parse_banner
#
# Parse the initial response banner the server sends when we connect.
# Hoping for:
#      220 blah blah <auth.mime> <msgid>
# The <auth.mime> string gives a list of supported extensions.
# The last bit is a msg-id, which identifies this connection,
# and is used in authentication, for example.
#
#=======================================================================
sub _parse_banner
{
    my $self   = shift;
    my $banner = shift;
    my ($code, $capstring, $msgid);


    ${*$self}{'net_dict_banner'} = $banner;
    ${*$self}{'net_dict_capabilities'} = [];
    if ($banner =~ /^(\d{3}) (.*) (<[^<>]*>)?\s+(<[^<>]+>)\s*$/)
    {
        ${*$self}{'net_dict_msgid'} = $4;
        ($capstring = $3) =~ s/[<>]//g;
        if (length($capstring) > 0)
        {
            ${*$self}{'net_dict_capabilities'} = [split(/\./, $capstring)];
        }
    }
    else
    {
        carp "unexpected format for welcome banner on connection:\n",
             $banner if $self->debug;
    }
}

#=======================================================================
#
# _get_database_list
#
# Get the list of databases on the remote server.
# We cache them in the instance data object, so that dbTitle()
# and databases() don't have to go to the server every time.
#
# We check to see whether we've already got the databases first,
# and do nothing if so. This means that this private method
# can just be invoked in the public methods.
# 
#=======================================================================
sub _get_database_list
{
    my $self = shift;


    return if exists ${*$self}{'net_dict_dbs'};

    if ($self->_SHOW_DB)
    {
	my($dbNum)= ($self->message =~ /^\d{3} (\d+)/);
	my($name, $descr);
 	foreach (0..$dbNum-1)
        {
            ($name, $descr) = (split /\s/, $self->getline, 2);
            chomp $descr;
            ${${*$self}{'net_dict_dbs'}}{$name} = _unquote($descr);
	}
	# Is there a way to do it right? Reading the dot line and the
	# status line afterwards? Maybe I should use read_until_dot?
	$self->getline();
	$self->getline();
    }
}

#-----------------------------------------------------------------------
# Method aliases for backwards compatibility
#-----------------------------------------------------------------------
*strats = \&strategies;

1;



Makefile.PL
#
# Makefile.PL for Net-Dict
#
# $Id: Makefile.PL,v 1.2 2003/05/05 23:56:17 neilb Exp $
#

use ExtUtils::MakeMaker;

my $DEFAULT_HOST = 'dict.org';
my $DEFAULT_PORT = 2628;

$test_config_file = 't/test_host.cfg';

print <<EOF;

			Net::Dict

EOF

&configure_testsuite();

&WriteMakefile(
                NAME		=> 'Net::Dict',
                DISTNAME	=> 'Net-Dict',
            VERSION_FROM	=> 'lib/Net/Dict.pm',
                PREREQ_PM	=> {
                                 'IO::Socket'     => 0,
                                 'Net::Cmd'	      => 0,
                                 'Carp'		      => 0,
                                 'AppConfig::Std' => 0,
                                },
		EXE_FILES	=> [qw(dict tkdict)],
		AUTHOR		=> 'Neil Bowers <neil@bowers.com>',
		ABSTRACT_FROM	=> 'lib/Net/Dict.pod',
        META_MERGE => {
            resources => {
                repository => 'https://github.com/neilbowers/Net-Dict',
            },
        },
        LICENSE      => 'perl',

                dist		=> {COMPRESS => 'gzip', SUFFIX => 'gz'},
              );

sub configure_testsuite
{
    my ($host, $port);

print <<EOF;
Testsuite Configuration

The testsuite uses the DICT server at dict.org.
If you're going to run the testsuite, then please confirm
that the hostname and port are correct. If you're going
to skip "make test", then just hit return on the two questions.

EOF

    $host = prompt('what is the hostname for the DICT server?', $DEFAULT_HOST);
    $port = prompt('what is the port for the DICT server?', $DEFAULT_PORT);
    print "\nWriting test config $test_config_file\n\n";

    open(CFGFILE, "> $test_config_file")
	|| die "failed to write $test_config_file: $!\n";

    print CFGFILE <<EOF;
#
# Testsuite configuration
#
# This file is auto-generated by Makefile.PL
#
use vars qw(\$HOST \$PORT);
\$HOST = '$host';
\$PORT = $port;
1;
EOF

    close(CFGFILE);
}

